jerry-core/ecma/builtin-objects/ecma-builtin-object.c

```diff
@@ -1329,6 +1303,83 @@ ecma_builtin_object_dispatch_routine (uint16_t builtin_routine_id, /**< built-in
 #endif /* ENABLED (JERRY_ESNEXT) */
     return result;
   }
+  else if (builtin_routine_id <= ECMA_OBJECT_ROUTINE_GET_OWN_PROPERTY_SYMBOLS)
+  {
+#if ENABLED (JERRY_ESNEXT)
+    ecma_value_t object = ecma_op_to_object (arg1);
+
+    if (ECMA_IS_VALUE_ERROR (object))
+    {
+      return object;
+    }
+
+    obj_p = ecma_get_object_from_value (object);
+
+    ecma_collection_t *props_p = ecma_op_object_own_property_keys (obj_p);
+
+    if (props_p == NULL)
+    {
+      ecma_deref_object (obj_p);
+      return ECMA_VALUE_ERROR;
+    }
+
+    if (props_p->item_count == 0)
+    {
+      ecma_deref_object (obj_p);
+      ecma_collection_destroy (props_p);
+      return ecma_op_create_array_object (NULL, 0, false);
+    }
+
+    ecma_collection_t *symbol_props = ecma_new_collection ();
+    ecma_collection_t *non_symbol_props = ecma_new_collection ();
+
+    for (uint32_t i = 0; i < props_p->item_count; i++)
+    {
+      ecma_value_t prop_name = props_p->buffer_p[i];
+      ecma_string_t *name_p = ecma_get_prop_name_from_value (prop_name);
+
+      ecma_ref_ecma_string (name_p);
+
+      if (ecma_prop_name_is_symbol (name_p))
+      {
+        ecma_collection_push_back (symbol_props, prop_name);
+      }
+      else
+      {
+        ecma_collection_push_back (non_symbol_props, prop_name);
+      }
+    }
+
+    ecma_value_t result_array;
+
+    if (builtin_routine_id == ECMA_OBJECT_ROUTINE_GET_OWN_PROPERTY_SYMBOLS)
+    {
+      result_array = ecma_op_create_array_object (symbol_props->buffer_p, symbol_props->item_count, false);
+    }
+    else
+    {
+      result_array = ecma_op_create_array_object (non_symbol_props->buffer_p, non_symbol_props->item_count, false);
+    }
+
+    ecma_deref_object (obj_p);
+    ecma_collection_free (symbol_props);
+    ecma_collection_free (non_symbol_props);
+#else /* !ENABLED (JERRY_ESNEXT) */
+    obj_p = ecma_get_object_from_value (arg1);
+    ecma_collection_t *props_p = ecma_op_object_own_property_keys (obj_p);
+
+    if (props_p->item_count == 0)
+    {
+      ecma_collection_destroy (props_p);
+      return ecma_op_create_array_object (NULL, 0, false);
+    }
+
+    ecma_value_t result_array = ecma_op_create_array_object (props_p->buffer_p, props_p->item_count, false);
```

Comment:
This is a fairly big chunk of code. We should move it into a method.

Reply:
+1